docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+* Enabling the creation of user interfaces that can guide an application deployer through the deployment process with descriptive help about the configuration value decisions they are making, and useful default values where appropriate
+
+### Use cases for parameters within templates
+
+* Generate and share passwords between components (parameter value is provided to each component as an environment variable)
+* Allow for simple deployment-time customization of “app” configuration via environment values or api objects, e.g. memory 
+  tuning parameters to a MySQL image, Docker image registry prefix for image strings, pod resource requests and limits, default 
+  scale size.
+* Allow simple, declarative defaulting of parameter values AND expose them to end users in an approachable way - a parameter 
+  like “MySQL table space” can be parameterized in images as an env var - the template parameters declare the parameter, give 
+  it a friendly name, give it a reasonable default, and informs the user what tuning options available.
+* Customization of component names to avoid collisions and ensure matched labeling (eg replica selector value and pod label are 
+  user provided and in sync).
+* Customize cross-component references (eg user provides the name of a secret that already exists in their namespace, to use in 
+  a pod as a TLS cert).
+* Provide guidance to users for parameters such as default values, descriptions, and whether or not a particular parameter value 
+  is required or can be left blank.
+* Parameterize the replica count of a deployment or PetSet
+* Parameterize part of the labels and selector for a DaemonSet
+
+
+## Design Assumptions
+
+The goal for this proposal is a simple schema which addresses a few basic challenges:
+
+* Allow application authors to expose configuration knobs for application deployers, with suggested defaults and 
+descriptions of the purpose of each knob
+* Allow application deployers to easily customize exposed values like object names while maintaining referential integrity 
+  between dependent pieces
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
+This desire for simplicity also intentionally excludes template composability/embedding as a supported use case.  
+
```

Comment:
including things by reference leads to fragility as the things you reference bit rot or disappear or get changed underneath you.  Furthermore a consumer of something that includes 3 other things by reference, if they are diligent, is going to need to go track down those other 3 things to understand what they are going to do before blindly instantiating it.

obviously i'm not saying there's no place for DRY, but in my experience writing templates so far, i think embedding them would have only made my life harder in trying to keep everything in sync when i changed something that other templates depended on, rather than making it simpler by allowing to me write something one time.  And that's in a scenario where i had total control over the entire hierarchy of template content.

lastly the biggest challenge we've had with templates so far is getting people to understand how to write them (what's the valid syntax, what's the schema for the objects, etc), and i'm worried that adding additional semantics like includes is only going to raise the bar on understanding them.

Regardless i'd like to see some other folks weigh in before we remove the restriction, it's a pretty clear decision point so hopefully we can get some consensus one way or another as we attempt to define the proposed syntax.